wfe2/wfe.go

```diff
@@ -2377,6 +2378,71 @@ func (wfe *WebFrontEndImpl) RenewalInfo(ctx context.Context, logEvent *web.Reque
 		time.Unix(0, cert.Expires).UTC()))
 }
 
+// UpdateRenewal is used by the client to inform the server that they have
+// replaced the certificate in question, so it can be safely revoked. All
+// requests must be authenticated to the account which ordered the cert.
+func (wfe *WebFrontEndImpl) UpdateRenewal(ctx context.Context, logEvent *web.RequestEvent, response http.ResponseWriter, request *http.Request) {
+	if !features.Enabled(features.ServeRenewalInfo) {
+		wfe.sendError(response, logEvent, probs.NotFound("Feature not enabled"), nil)
+		return
+	}
+
+	body, _, acct, prob := wfe.validPOSTForAccount(request, ctx, logEvent)
+	addRequesterHeader(response, logEvent.Requester)
+	if prob != nil {
+		// validPOSTForAccount handles its own setting of logEvent.Errors
+		wfe.sendError(response, logEvent, prob, nil)
+		return
+	}
+
+	var updateRenewalRequest struct {
+		CertID   string `json:"certID"`
+		Replaced bool   `json:"replaced"`
+	}
+	err := json.Unmarshal(body, &updateRenewalRequest)
+	if err != nil {
+		wfe.sendError(response, logEvent, probs.Malformed("Unable to unmarshal RenewalInfo POST request body"), err)
+		return
+	}
+
+	der, err := base64.RawURLEncoding.DecodeString(updateRenewalRequest.CertID)
+	if err != nil {
+		wfe.sendError(response, logEvent, probs.Malformed("certID was not base64url-encoded"), nil)
+		return
+	}
+
+	var id certID
```

Comment:
The following chunk of code is used almost verbatim (barring a comment) in both the `RenewalInfo` and `UpdateRenewal` methods. Could you create a helper function that performs these checks and call the helper from each method instead?
```
	var id certID
	rest, err := asn1.Unmarshal(der, &id)
	if err != nil || len(rest) != 0 {
		wfe.sendError(response, logEvent, probs.Malformed("Path was not a DER-encoded CertID sequence"), nil)
		return
	}

	// Verify that the hash algorithm is SHA-256, so people don't use SHA-1 here.
	if !id.HashAlgorithm.Algorithm.Equal(asn1.ObjectIdentifier{2, 16, 840, 1, 101, 3, 4, 2, 1}) {
		wfe.sendError(response, logEvent, probs.Malformed("Request used hash algorithm other than SHA-256"), nil)
		return
	}
```

Reply:
I could, and in most other files I would. But doing that sort of helper-function extraction is actually a little bit awkward in the WFE because we don't use "normal" error handling here. In each of these error paths, we call `wfe.sendError()` and pass it _both_ the Problem string that we want the user to see, and the actual underlying error for our logs.

Creating a helper function for processing the certID would look something like this:
```go
func serialFromCertID(certID string) (serial string, prob *probs.ProblemDetails, err error) {
	der, err := base64.RawURLEncoding.DecodeString(certID)
	if err != nil {
		return "", probs.Malformed("CertID was not base64url-encoded or had padding"), err
	}

	var id certID
	rest, err := asn1.Unmarshal(der, &id)
	if err != nil || len(rest) != 0 {
		return "", probs.Malformed("CertID was not a DER-encoded CertID sequence"), err
	}

	if !id.HashAlgorithm.Algorithm.Equal(asn1.ObjectIdentifier{2, 16, 840, 1, 101, 3, 4, 2, 1}) {
		return "", probs.Malformed("CertID used hash algorithm other than SHA-256"), err
	}

	return core.SerialToString(id.SerialNumber)
}
```
and then calling it would look like
```go
	serial, prob, err := serialFromCertID(updateRenewalRequest.CertID)
	if prob != nil || err != nil {
		wfe.sendError(response, logEvent, prob, err)
		return
	}
	logEvent.Extra["RequestedSerial"] = serial
	beeline.AddFieldToTrace(ctx, "request.serial", serial)
```

It's totally possible, but the dual prob+err returns makes it awkward and non-idiomatic in a way I don't love. Particularly because future changes could accidentally change the `serialFromCertID` function to break the "alwasy return both a prob and an err, or neither" contract that it would start with.

Totally happy to do it this way if you prefer, just let me know! Just wanted to explain why I did it this way the first time.

Reply:
Ah yeah, that does make it a bit gross. Thank you for elaborating. I'm fine to leave it as-is now.